File: code-templates/python/optimization/pso_template.py
"""
粒子群优化 (PSO) 模板
适用：连续变量非线性优化
问题适配点：
  1. 修改 objective() 和 constraint_penalty() 方法
  2. 修改 bounds
  3. 调整 PSO 参数
"""

import numpy as np


class PSO:
    def __init__(self, obj_func, n_particles=30, max_iter=200,
                 w_start=0.9, w_end=0.4, c1=2.0, c2=2.0,
                 tol=1e-6, patience=50):
        self.obj = obj_func
        self.n_particles = n_particles
        self.max_iter = max_iter
        self.w_start = w_start
        self.w_end = w_end
        self.c1 = c1
        self.c2 = c2
        self.tol = tol
        self.patience = patience
        self.n_var = obj_func.n_var
        self.bounds = obj_func.bounds
        self.history = {'best_fitness': [], 'avg_fitness': []}

    def _init_swarm(self):
        positions = np.zeros((self.n_particles, self.n_var))
        velocities = np.zeros((self.n_particles, self.n_var))
        for i in range(self.n_var):
            low, high = self.bounds[i]
            positions[:, i] = np.random.uniform(low, high, self.n_particles)
            velocities[:, i] = np.random.uniform(-abs(high - low) * 0.1,
                                                  abs(high - low) * 0.1,
                                                  self.n_particles)
        return positions, velocities

    def _evaluate(self, positions):
        return np.array([self.obj.fitness(p) for p in positions])

    def run(self, verbose=True):
        positions, velocities = self._init_swarm()
        fitness = self._evaluate(positions)

        pbest_positions = positions.copy()
        pbest_fitness = fitness.copy()

        gbest_idx = np.argmax(fitness)
        gbest_position = positions[gbest_idx].copy()
        gbest_fitness = fitness[gbest_idx]

        no_improve = 0

        for it in range(self.max_iter):
            w = self.w_start - (self.w_start - self.w_end) * it / self.max_iter

            r1 = np.random.random((self.n_particles, self.n_var))
            r2 = np.random.random((self.n_particles, self.n_var))

            velocities = (w * velocities +
                          self.c1 * r1 * (pbest_positions - positions) +
                          self.c2 * r2 * (gbest_position - positions))

            # 速度限制
            for i in range(self.n_var):
                v_max = abs(self.bounds[i][1] - self.bounds[i][0]) * 0.2
                velocities[:, i] = np.clip(velocities[:, i], -v_max, v_max)

            positions = positions + velocities

            # 位置边界处理
            for i in range(self.n_var):
                low, high = self.bounds[i]
                positions[:, i] = np.clip(positions[:, i], low, high)

            fitness = self._evaluate(positions)

            # 更新个体最优
            improved = fitness > pbest_fitness
            pbest_positions[improved] = positions[improved]
            pbest_fitness[improved] = fitness[improved]

            # 更新全局最优
            if np.max(fitness) > gbest_fitness + self.tol:
                gbest_idx = np.argmax(fitness)
                gbest_position = positions[gbest_idx].copy()
                gbest_fitness = fitness[gbest_idx]
                no_improve = 0
            else:
                no_improve += 1

            self.history['best_fitness'].append(gbest_fitness)
            self.history['avg_fitness'].append(np.mean(fitness))

            if verbose and it % 20 == 0:
                print(f"Iter {it:4d} | Best: {gbest_fitness:.6f}")

            if no_improve >= self.patience:
                if verbose:
                    print(f"收敛于第 {it} 次迭代")
                break

        return gbest_position, gbest_fitness, self.history


if __name__ == "__main__":
    from ga_template import ObjectiveFunction
    obj = ObjectiveFunction()
    pso = PSO(obj, n_particles=30, max_iter=200)
    solution, fitness, history = pso.run()
    print(f"\n最优解: {solution}")
    print(f"最优适应度: {fitness:.6f}")

File: code-templates/matlab/optimization/pso_template.m
% 粒子群优化 (PSO) MATLAB 模板

function [gbest_x, gbest_f, history] = pso_template()
    % ===== 参数设置 =====
    n_particles = 30;
    n_var = 2;
    bounds = [0, 10; 0, 10];
    max_iter = 200;
    w_start = 0.9; w_end = 0.4;
    c1 = 2.0; c2 = 2.0;
    patience = 50;
    tol = 1e-6;

    % ===== 初始化 =====
    pos = zeros(n_particles, n_var);
    vel = zeros(n_particles, n_var);
    for i = 1:n_var
        pos(:, i) = bounds(i,1) + (bounds(i,2)-bounds(i,1)) * rand(n_particles, 1);
        vel(:, i) = (bounds(i,2)-bounds(i,1)) * 0.1 * randn(n_particles, 1);
    end

    fitness = evaluate_pop(pos);

    pbest_pos = pos;
    pbest_fit = fitness;
    [gbest_f, idx] = max(fitness);
    gbest_x = pos(idx, :);
    no_improve = 0;

    history.best_fitness = zeros(max_iter, 1);
    history.avg_fitness = zeros(max_iter, 1);

    % ===== 主循环 =====
    for it = 1:max_iter
        w = w_start - (w_start - w_end) * it / max_iter;

        r1 = rand(n_particles, n_var);
        r2 = rand(n_particles, n_var);

        vel = w * vel + c1 * r1 .* (pbest_pos - pos) + c2 * r2 .* (gbest_x - pos);

        % 速度限制
        for i = 1:n_var
            v_max = abs(bounds(i,2) - bounds(i,1)) * 0.2;
            vel(:, i) = max(-v_max, min(v_max, vel(:, i)));
        end

        pos = pos + vel;

        % 边界处理
        for i = 1:n_var
            pos(:, i) = max(bounds(i,1), min(bounds(i,2), pos(:, i)));
        end

        fitness = evaluate_pop(pos);

        improved = fitness > pbest_fit;
        pbest_pos(improved, :) = pos(improved, :);
        pbest_fit(improved) = fitness(improved);

        [iter_best, idx] = max(fitness);
        if iter_best > gbest_f + tol
            gbest_f = iter_best;
            gbest_x = pos(idx, :);
            no_improve = 0;
        else
            no_improve = no_improve + 1;
        end

        history.best_fitness(it) = gbest_f;
        history.avg_fitness(it) = mean(fitness);

        if mod(it, 20) == 0
            fprintf('Iter %4d | Best: %.6f\n', it, gbest_f);
        end

        if no_improve >= patience
            fprintf('收敛于第 %d 次迭代\n', it);
            break;
        end
    end

    figure;
    plot(history.best_fitness(1:it), 'b-', 'LineWidth', 1.5); hold on;
    plot(history.avg_fitness(1:it), 'r-', 'LineWidth', 1);
    xlabel('Iteration'); ylabel('Fitness');
    legend('Best', 'Average');
    title('PSO Convergence');
end

function fit = evaluate_pop(pos)
    [n, ~] = size(pos);
    fit = zeros(n, 1);
    for i = 1:n
        fit(i) = obj_func(pos(i, :));
    end
end

function f = obj_func(x)
    % TODO: 替换为实际目标函数
    f = -sum(x.^2);
end
